bin/proving-service/src/proxy/worker.rs

```diff
@@ -1,75 +1,216 @@
+            return Some(WorkerHealthStatus::Unhealthy {
+                failed_attempts: failed_attempts + 1,
+                first_fail_timestamp: Instant::now(),
+                reason: "Unsupported proof type".to_string(),
+            });
+        }
+
+        if !worker_status.ready {
+            return Some(WorkerHealthStatus::Unhealthy {
+                failed_attempts: failed_attempts + 1,
+                first_fail_timestamp: Instant::now(),
+                reason: "Worker not ready".to_string(),
+            });
         }
+
+        Some(WorkerHealthStatus::Healthy)
     }
 
+    /// Returns the worker availability.
     pub fn is_available(&self) -> bool {
         self.is_available
     }
 
+    /// Sets the worker availability.
     pub fn set_availability(&mut self, is_available: bool) {
-        self.is_available = is_available;
+        self.is_available = is_available
+    }
+
+    pub(crate) fn set_health_status(&mut self, health_status: WorkerHealthStatus) {
+        self.health_status = health_status;
+        match &self.health_status {
+            WorkerHealthStatus::Healthy => {
+                self.is_available = true;
+            },
+            WorkerHealthStatus::Unhealthy { .. } => {
+                WORKER_UNHEALTHY.with_label_values(&[&self.address()]).inc();
+                self.is_available = false;
+            },
+        }
+    }
+
+    /// Returns the number of retries the worker has had.
+    pub fn retries_amount(&self) -> usize {
+        match &self.health_status {
+            WorkerHealthStatus::Healthy => 0,
+            WorkerHealthStatus::Unhealthy {
+                failed_attempts,
+                first_fail_timestamp: _,
+                reason: _,
+            } => *failed_attempts,
+        }
+    }
+
+    /// Returns whether the worker should do a health check.
+    ///
+    /// A worker should do a health check if it is healthy or if the time since the first failure
+    /// is greater than the time since the first failure power of 2.
+    ///
+    /// The maximum exponent is [MAX_BACKOFF_EXPONENT], which corresponds to a backoff of
+    /// 2^[MAX_BACKOFF_EXPONENT] seconds.
+    pub(crate) fn should_do_health_check(&self) -> bool {
+        match self.health_status {
+            WorkerHealthStatus::Healthy => true,
+            WorkerHealthStatus::Unhealthy {
+                failed_attempts,
+                first_fail_timestamp,
+                reason: _,
+            } => {
+                let time_since_first_failure = Instant::now() - first_fail_timestamp;
+                time_since_first_failure
+                    > Duration::from_secs(
+                        2u64.pow(failed_attempts.min(MAX_BACKOFF_EXPONENT) as u32),
+                    )
+            },
+        }
+    }
+
+    pub fn health_status(&self) -> &WorkerHealthStatus {
+        &self.health_status
+    }
+
+    pub fn version(&self) -> &str {
+        &self.version
     }
```

Comment:
These should probably go into the "PUBLIC ACCESSORS" section. In general, lets better organize this file to have our usual "CONSTRUCTORS", "PUBLIC ACCESSORS", "STATE MUTATORS" etc. sections.

I'd also add doc comments to these methods.